Remember blame splitter position

mdonatas opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Fixes #12225

## Proposed changes

- Save/remember BlameControl's splitter positions

One caveat is that without the user adjusting splitter's position at least once, the "broken" state will remain. I haven't found an easy and sensible way to set some kind of default for the splitter.

## Test methodology <!-- How did you ensure quality? -->

- Manual. I've been running custom build with this commit for a little bit over 1 year and didn't have a problem with Blame ever since.

## Test environment(s) <!-- Remove any that don't apply -->

- Windows 11 200% scaling

<!-- Mention language, UI scaling, or anything else that might be relevant -->

## Merge strategy

<!-- Change the following if the merge strategy should be changed:
- Squash merge (maintainer to decide merge message, PR submitter should cleanup commits/messages at PR approval).
- Rebase merge (PR submitter must change the commit message for the last commit).
- Merge commit. (PR submitter to rebase and squash before merges).
- To be decided later.
The maintainer may still request the contributor to squash and rebase, to make sure that merges and commit messages are clarified.
-->

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

vbjay commented:
Detect less than x value for splitter width or height for each splitter and reset to reasonable value if detected. That way app auto fixes to at least show both sides. Make x small so that it doesn't reset when user wants it small. But example doesn't make sense to be 20 pixels wide for the code side. 

OR
Add menu to view menu that resets splitters to middle either width or height if vertical or horizontal. Then let user hit menu to autofit all splitters.

mstv commented:
Did you want to comment to the orthogonal improvement by #12250, @vbjay?

mstv commented on src/app/GitUI/UserControls/BlameControl.cs:
```diff
@@ -82,6 +82,12 @@ public BlameControl()
             _gitBlameParser = new GitBlameParser(() => UICommands.Module);
         }
 
+        public void InitSplitterManager(SplitterManager splitterManager, string parentContainer)
+        {
+            splitterManager.AddSplitter(splitContainer1, $"{parentContainer}.BlameSplitContainer1");
+            splitterManager.AddSplitter(splitContainer2, $"{parentContainer}.BlameSplitContainer2");
```
```suggestion
        public void InitSplitterManager(SplitterManager splitterManager, string parentContainerPrefix)
        {
            string containerPrefix = $"{parentContainerPrefix}{Name}.";     
            splitterManager.AddSplitter(splitContainer1, $"{containerPrefix }{nameof(splitContainer1)}");
            splitterManager.AddSplitter(splitContainer2, $"{containerPrefix }{nameof(splitContainer1)}");
```

mstv commented on src/app/GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -163,6 +163,7 @@ public void ExpandToFile(string filePath, int? line, bool requestBlame)
         public void InitSplitterManager(SplitterManager splitterManager)
         {
             splitterManager.AddSplitter(FileTreeSplitContainer, "FileTreeSplitContainer");
+            BlameControl.InitSplitterManager(splitterManager, "RevisionFileTree");
```
There is no need to adapt this because the `RevisionFileTreeControl` class will be deleted when it will be replaced with `RevisionDiffControl` in the near future.

mdonatas replied:
Doesn't hurt though :)

vbjay commented:
> Did you want to comment to the orthogonal improvement by #12250, @vbjay?

Yes. Sorry. Just saw this pr.

mdonatas commented:
Addressed comments. Commit `Use parent container Names rather than custom values`
Also, added default value initialization, which really helps!

Here are the screenshots with 100% and 200% scaling when opening GE with splitter settings not yet saved

`100%`
![image](https://github.com/user-attachments/assets/10431bd1-4f3c-4160-9014-809a85889ea0)

`200%`
![image](https://github.com/user-attachments/assets/0fe5d6a2-1ac2-4081-a587-3b12405ae218)

mstv commented on src/app/GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -346,6 +346,7 @@ public void InitSplitterManager(SplitterManager splitterManager)
         {
             splitterManager.AddSplitter(DiffSplitContainer, nameof(DiffSplitContainer));
             splitterManager.AddSplitter(LeftSplitContainer, $"{nameof(RevisionDiffControl)}.{nameof(LeftSplitContainer)}");
+            BlameControl.InitSplitterManager(splitterManager, parentContainerPrefix: Name);
```
`parentContainerPrefix: Name` may work for this PR. But it is not exactly what is meant. The _prefix_ shall contain the trailing dot - in order to apply this as generic pattern in a follow-up PR.

mstv commented on src/app/GitUI/SplitterManager.cs:
```diff
@@ -99,6 +99,10 @@ public void RestoreFromSettings(SettingsSource settings)
                 }
+                else if (_defaultDistance.HasValue)
+                {
+                    SetSplitterDistance(_defaultDistance.Value);
```
I think this change will make it impossible to persist the collapsed state, e.g. of the Left Panel IIRC.

mdonatas replied:
I've changed this to also check for whether there hasn't been a setting saved before. By the way, this overload is not used anywhere else at the moment.

mstv commented on src/app/GitUI/UserControls/BlameControl.cs:
```diff
@@ -6,6 +6,7 @@
 using GitExtensions.Extensibility.Git;
 using GitExtensions.Extensibility.Plugins;
 using GitExtUtils;
+using GitExtUtils.GitUI;
```
MSVS was hyperactive most likely.
```suggestion
```

mstv replied:
superfluous addition

mdonatas replied:
Oops, removed

mstv commented on src/app/GitUI/UserControls/BlameControl.cs:
```diff
@@ -82,6 +83,13 @@ public BlameControl()
+        public void InitSplitterManager(SplitterManager splitterManager, string parentContainerPrefix)
+        {
+            string containerPrefix = $"{parentContainerPrefix}{Name}.";
+            splitterManager.AddSplitter(splitContainer1, $"{containerPrefix}{splitContainer1.Name}", defaultDistance: splitContainer1.Panel1MinSize + 1);
+            splitterManager.AddSplitter(splitContainer2, $"{containerPrefix}{splitContainer2.Name}", defaultDistance: splitContainer2.Panel1MinSize + 1);
```
What is the clue behind `+ 1`?

mdonatas replied:
There is a validation in a `SplitterManager` where this has to be true. https://github.com/mdonatas/gitextensions/blob/remember-blame-splitter-pos/src/app/GitUI/SplitterManager.cs#L147-L150

I suppose the logic is that the splitter can't occupy the same position as `panel1` at it's smallest size as that would require it to get even smaller to leave space for the splitter and it can't.

mstv reviewed (commented):
👍
final nits

mstv commented:
`verified` files need an update yet.

mdonatas commented:
Added several tests for SplitterManager to cover default value bahavior

mstv commented:
I have squashed, resolved the MC and made use of `NestedSplitterManager`: mstv/remember-blame-splitter-pos: 00ee78a5ab8b2ce00d9c8638905cd41e2b5bea33.
mstv/remember-blame-splitter-pos-before_squash: 77cd75eb46be6336fa804d45066aa18d9fc3d8ec
mstv/remember-blame-splitter-pos-squashed: 90ce5920be3c1e224bd22a82b7193fa47eb43fbb

mstv reviewed (approved):
Thank you
